sources/en/n/novelsemperor.py

```diff
@@ -17,3 +18,85 @@ def initialize(self) -> None:
                 "promotion, charge 100 and get 500 VIP coupons!",
             ]
         )
+
+    def search_novel(self, query):
+        query = query.lower().replace(" ", "+")
+        soup = self.get_soup(search_url % query)
+
+        results = []
+        for tab in soup.select("div.xlg\\:grid-cols-8.grid.grid-cols-3 div#card-real"):
+            a = tab.select_one("a")
+            title = a["href"][33:].replace('-', ' ')
+            img = tab.select_one("img")["data-src"]
+            results.append(
+                {
+                    "title": title,
+                    "url": self.absolute_url(a["href"]),
+                    "img": img,
+                }
+            )
+
+        return results
+
+    def read_novel_info(self):
+        logger.debug("Visiting %s", self.novel_url)
+        pagination_num = self.get_soup(self.novel_url)
+        pagination_num = pagination_num.select_one("ul > li:nth-last-child(2)").text
+        pagination_num = int(pagination_num)
+
+        soup = []
+
+        for i in range(1, pagination_num + 1):
+            soup.append(self.get_soup(f'{self.novel_url}?page={i}'))
```

Comment:
this will fetch pages one by one. it can slow down the crawler signigicantly. you can use parallel executors. example:
https://github.com/dipu-bd/lightnovel-crawler/blob/431c3687871e4da99b4326f2774c70dcd7746244/sources/zh/uukanshu.py#L62-L69

Reply:
Thanks for the review. I am new to python so thanks for the advice I'll look into the example.